### PR TITLE
add dml softmax_cross_entropy_with_logits kernel

### PR DESCRIPTION
**Pull request by zhangxiang1993**

### CODE REVIEW

File: tensorflow/core/common_runtime/dml/dml_operator_helper.h [hunk 1]
```diff
@@ -144,6 +144,22 @@ class BatchNormGradShapeHelper : public ShapeHelper {
       const InitializationHelper* initialization_helper) const override;
 };
 
+class SoftmaxXentWithLogitsShapeHelper : public ShapeHelper {
```
Comment (PatriceVignola) [inline tensorflow/core/common_runtime/dml/dml_operator_helper.h]: Don't add new shape helpers in this file. Only shape helpers that can be used by many kernels should be here. The BatchNorm helpers are relics from the past that should be moved out eventually. #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc:64]: Where is the validation? You should have similar validation to what the CPU kernel does.

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    // log(sum(exp(logits - max_logits)))
+    auto log_sum_exp = dml::Log(sum_exp);
+    auto log_sum_exp_bcast =
+        dml::Reinterpret(log_sum_exp, input_sizes, broadcast_c_strides);
+
+    // sum(-labels *
+    //    ((logits - max_logits) - log(sum(exp(logits - max_logits)))))
+    auto sub = shifted_logits - log_sum_exp_bcast;
+    auto mul = -1.0f * labels * sub;
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: you can simply do -labels here #Resolved

Comment (PatriceVignola) [reply tensorflow/core/kernels/dml_softmax_ops.cc]: Actually, scratch that. You can remove an operation by doing `sub = log_sum_exp_bcast - shifted_logits` instead. After all, `-1 * (x - y) = y - x` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    return {outputShape, logits.shape()};
+  }
+};
+
+class DmlSoftmaxXentWithLogitsInitHelper : public InitializationHelper {
+ public:
+  using Attributes = EmptyAttributes;
+
+  DmlSoftmaxXentWithLogitsInitHelper(OpKernelContext* ctx,
+                           std::shared_ptr<const Attributes> attr) {
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: alignment #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                                        "2-dimensional, or broadcasted to be "
+                                        "2-dimensional"));
+  }
+};
+
+class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
+ public:
+  using InitHelper = DmlSoftmaxXentWithLogitsInitHelper;
+  explicit DmlSoftmaxXentWithLogitsKernel(
+      DmlKernelConstruction* ctx, const DmlSoftmaxXentWithLogitsInitHelper* init_helper) {
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: line length #Resolved

**Review by PatriceVignola — APPROVED** [review]
:shipit:

**Comment by zhangxiang1993** [comment]
Should I hold this PR until directml-dev branch is created? @jstoecker

**Comment by jstoecker** [comment]
> Should I hold this PR until directml-dev branch is created? @jstoecker

Let's merge it.
